data_source.go

```diff
@@ -99,40 +120,90 @@ func Start(ds DataSource) error {
-	go func() {
-		defer ds.RunDoneDone()
-		for {
-			// fmt.Println("calling blockingRead")
-			if err := ds.blockingRead(); err == io.EOF {
-				log.Println("blockingRead returns io.EOF, more stopping the source")
-				// EOF error should occur after abortSelf has been closed
-				// ds.CloseOutputs() // why is this here, ds.Stop also calls CloseOutputs
+	go CoreLoop(ds, queuedRequests)
+	return nil
+}
+
+// CoreLoop has the DataSource produce data until graceful stop.
+// This will be a long-running goroutine, as long as a source is active.
+func CoreLoop(ds DataSource, queuedRequests chan func()) {
+	defer ds.RunDoneDeactivate()
+	nextBlock := ds.getNextBlock()
+
+	for {
+		// Use select to interleave 2 activities that should NOT be done concurrently:
+		// 1. Handle RPC requests to chage data processing parameters (e.g. trigger)
+		// 2. Handle new data and processes it
+		select {
+
+		// Handle RPC requests
+		case request := <-queuedRequests:
+			request()
+
+		// Handle data, or recognize the end of data
+		case block, ok := <-nextBlock:
+			if !ok {
+				// nextBlock was closed in the data production loop when abortSelf was closed
+				log.Println("nextBlock channel was closed; stopping the source normally")
+				return
+
+			} else if block.err != nil {
+				// errors in block indicate a problem with source: need to close down
+				log.Printf("nextBlock receives Error; stopping source: %s\n", block.err.Error())
 				return
-			} else if err != nil {
-				// other errors indicate a problem with source, need to close down
-				log.Printf("blockingRead returns Error, stopping source: %s\n", err.Error())
-				ds.Stop() // will cause next call to ds.blockingRead to return io.EOF
 			}
-			if ds.Running() { // process data segments if source is still running
-				// fmt.Println("ProcessSegments")
-				if err := ds.ProcessSegments(); err != nil {
-					log.Printf("processSegments returns Error, stopping source: %s\n", err.Error())
-					ds.Stop()
-				}
+			if err := ds.ProcessSegments(block); err != nil {
+				log.Printf("AnySource.ProcessSegments returns Error; stopping source: %s\n", err.Error())
+				return
 			}
+			// In some sources, ds.getNextBlock has to be called again to initiate the next
+			// data acquisition step (Lancero specifically).
+			nextBlock = ds.getNextBlock()
 		}
+	}
+}
+
+// Stop tells the data supply to deactivate.
+func (ds *AnySource) Stop() error {
+	ds.runMutex.Lock()
+	switch ds.sourceState {
+	case Inactive:
+		ds.runMutex.Unlock()
+		return fmt.Errorf("AnySource not active, cannot stop")
+
+	case Starting:
+		fmt.Println("deleteme: called Stop on a Starting source; how to handle this??")
+
+	case Active:
+		// This is the normal case: Stop on an Active source
 
-	}()
+	case Stopping:
+		ds.runMutex.Unlock()
+		fmt.Println("deleteme: called Stop again on a Stopping source")
+		return nil
+	}
+	ds.sourceState = Stopping
+	closeIfOpen(ds.abortSelf)
+	ds.runMutex.Unlock()
 
+	ds.RunDoneWait()
 	return nil
 }
 
```

Comment:
Seems like `closeIfOpen` should only be neccesary if we're worried about accidentally closing a channel 2x? Can we at least put a println on the case that doesn't close to see if it ever happens?